### ISSUE TITLE
Use pattern hints from converters when matching fields

### ISSUE THREAD

**Issue by vytas7** [falconry/falcon#1567]
At the time of writing, the `CompiledRouter` is first matching fields, and then subsequently using converters to validate/adapt them to parameters that are sent to a responder.

However, information such as allowed characters, desired length of a field etc is completely ignored when building routing code and/or regexps. In complex URI patterns with multiple fields in a single path segment, a user may expect these are correctly split, if, for instance, the first field was specified to be a 6-digit integer.

Consider the following example:
```
import falcon
import falcon.testing
import pytest


class DebugResource():
    def on_get(self, req, resp, **kwargs):
        resp.body = str(kwargs)
        resp.content_type = falcon.MEDIA_TEXT


@pytest.fixture
def client():
    api = falcon.API()
    api.add_route('/{yymmdd:int(6)}-{item_id:uuid}', DebugResource())
    return falcon.testing.TestClient(api)


def test_routing(client):
    resp = client.simulate_get('/190101-59559ee7-e661-43c8-9da1-0ab83b9e0d8e')
    assert resp.status_code == 200
```

Running `pytest` fails with `assert 404 == 200`, whereas the user expectation was that `yymmdd` was assigned with `190101`, and the UUID parameter with the remainder.

**Comment by vytas7** [comment]
I also added the **documentation** label -- even if we don't properly fix this in the foreseeable future, documentation should clearer reflect the current limitations.

**Comment by CaselIT** [comment]
The also fails `/r/{foo:int}-{bar}` to match `/r/42-foo-bar`
The issue is that the regexp created is `r'^(?P<foo>.+)-(?P<bar>.+)$'` and matches `{'foo': '42-foo', 'bar':'bar'}` instead of the correct `{'foo': '42', bar: 'foo-bar'}`
It may make sense to use instead `r'^(?P<foo>.+?)-(?P<bar>.+?)$'` this regexp, (it would potentially be breaking change, that we may make pass as a bugfix)

**Comment by vytas7** [comment]
IMHO changing to non-greedy matching would not improve anything, it's just a mirror problem.

If your pattern was `/path/{foo}-{bar:int}`. Using the current version, a path like `/path/foo-bar-42` would work correctly (`foo-bar` and `42` would be matched). Conversely, matching in a non-greedy fashion would introduce a regression.

**Comment by CaselIT** [comment]
Indeed, still if we implement this it may be an option to switch che default to nonl-greedy match
